### PR TITLE
Improve JSON

### LINKED ISSUE #9: Make JSON better
https://github.com/tfausak/octane/issues/6#issuecomment-224913489

### PR DESCRIPTION
**Pull request by tfausak**
This will fix #9. It's still a work in progress.

### CODE REVIEW

**Comment by tfausak** [comment]
The new JSON is way smaller.

| Description | Commit | Size (MB) | Size (bytes) |
| --- | --- | --- | --- |
| Replay | - | 4.3 | 4324393 |
| Original JSON | d4bd9f9e22c8e81a35d525a5af9266501876ee52 | 123.3 | 123326034 |
| Naive JSON | aabc161bd099f9883db3c516217c91bef6b8e0bf | 92.2 | 92153986 |
| Optimized JSON | 7e463ab633011cfdf5e57b2739a67a1175d69b0c | 8.3 | 8365220 |

File: library/Octane/FullReplay.hs [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+            & Text.pack
+        , replication
+            & Parser.replicationProperties
+            & Map.map (\ property -> Aeson.object
+                [ ("Type", getPropertyName property)
+                , ("Value", getPropertyValue property)
+                ])
+        ))
+    & Prelude.filter (\ (_, properties) ->
+        Map.null properties)
```
Comment (tfausak) [inline library/Octane/FullReplay.hs]: Oops! This is the opposite of what I want. That's why the JSON is so small: It doesn't have anything in it!

**Comment by tfausak** [comment]
After fixing that terrible typo that caused basically nothing to be replicated, the JSON is now about 40 MB (41638624 bytes). That's still a huge improvement over the naive approach — almost 50%.

**Comment by tfausak** [comment]
In addition to the 50% size reduction, these changes make Octane execute about 50% faster. I ran some tests piping the output to `/dev/null`:
-   d4bd9f9e22c8e81a35d525a5af9266501876ee52
  
  ```
  Parsed 4223.0 kb in 11.1 s at 380.1 kbps.
  Parsed 4223.0 kb in 11.1 s at 378.8 kbps.
  Parsed 4223.0 kb in 11.2 s at 378.6 kbps.
  ```
-   aabc161bd099f9883db3c516217c91bef6b8e0bf
  
  ```
  Parsed 4223.0 kb in 7.8 s at 538.7 kbps.
  Parsed 4223.0 kb in 7.8 s at 542.2 kbps.
  Parsed 4223.0 kb in 7.8 s at 542.7 kbps.
  ```
-   7e463ab633011cfdf5e57b2739a67a1175d69b0c
  
  ```
  Parsed 4223.0 kb in 5.1 s at 833.7 kbps.
  Parsed 4223.0 kb in 4.9 s at 854.8 kbps.
  Parsed 4223.0 kb in 5.0 s at 852.9 kbps.
  ```

Pretty much linear speedup with the output file size.

**Comment by tfausak** [comment]
For reference, this is roughly what the new JSON looks like:

``` json
{
    "Version": "123.456",
    "Metadata": {},
    "Levels": ["usually only one"],
    "Messages": {
        "123": "(key is frame number) only old replays have messages"
    },
    "TickMarks": {
        "123": "(key is frame number) either a goal or a save"
    },
    "Frames": [
        {
            "IsKeyFrame": true,
            "Number": 0,
            "Time": 12.34,
            "Delta": 0.05,
            "Spawned": {
                "123": {
                    "Name": "...",
                    "Class": "...",
                    "Position": [1, 2, 3],
                    "Rotation": null
                }
            },
            "Updated": {
                "123": {
                    "property name": {
                        "Type": "whatever",
                        "Value": "anything"
                    }
                }
            },
            "Destroyed": [
                123
            ]
        }
    ]
}
```

**Comment by danielsamuels** [comment]
Is there any chance of getting the class name back? I make use of it to detect whether a particular actor is a player or not.  Now it seems like I'll have to check to see whether the actor contains a property known to only appear on a player actor, this seems very clunky to me and I'd rather not have to do it.

**Comment by tfausak** [comment]
Aren't the class names still there? When an actor spawns, I include its name and its class name, along with the position and rotation.

**Comment by danielsamuels** [comment]
Ah, so the class isn't included in the `Updated` data?

**Comment by tfausak** [comment]
No, because it never changes.

**Comment by danielsamuels** [comment]
Ok, I guess I'll just have to maintain an Actor ID => Class map myself.

**Pull request merged**